File: lib/src/Cache.php
<?php
declare(strict_types=1);

namespace VaultCode;

/**
 * Cache locale delle content key — CIFRATA e AUTENTICATA (AES-256-GCM: il tag
 * fa anche da firma, impedisce cache poisoning). §5.
 *
 * File su disco = nonce(12) ‖ ciphertext‖tag. Il contenuto è una mappa
 * "module:ck" → {content_key_hex, window, fetched_at}. La freschezza (TTL) è
 * decisa dal chiamante (Runtime) confrontando fetched_at con now: questo
 * permette il fallback "stale entro il grace" se il server è irraggiungibile.
 */
final class Cache
{
    private const AAD = 'vaultcode/cache/v1';

    private string $path;
    private string $key; // 32 byte

    public function __construct(string $path, string $key)
    {
        $this->path = $path;
        $this->key = $key;
    }

    /** Deriva la chiave di cache dal segreto per-installazione (deterministica). */
    public static function deriveKey(string $installSecret): string
    {
        return substr(hash('sha256', $installSecret . 'vaultcode/cache-key/v1', true), 0, 32);
    }

    private function load(): array
    {
        if ($this->path === '' || !is_file($this->path)) { return []; }
        $blob = (string)file_get_contents($this->path);
        if (strlen($blob) < 13) { return []; }
        $nonce = substr($blob, 0, 12);
        $ct = substr($blob, 12);
        try {
            $json = Crypto::aesGcmDecrypt($this->key, $ct, $nonce, self::AAD);
        } catch (\Throwable $e) {
            return []; // cache corrotta/manomessa → ignorata (verrà riscritta)
        }
        $data = json_decode($json, true);
        return is_array($data) ? $data : [];
    }

    private function save(array $data): void
    {
        if ($this->path === '') { return; }
        $nonce = Crypto::newNonce();
        $ct = Crypto::aesGcmEncrypt($this->key, json_encode($data), $nonce, self::AAD);
        $dir = dirname($this->path);
        if (!is_dir($dir)) { @mkdir($dir, 0700, true); }
        file_put_contents($this->path, $nonce . $ct, LOCK_EX);
        @chmod($this->path, 0600);
    }

    /** @return array{content_key_hex:string,window:string,fetched_at:int}|null */
    public function get(string $module, int $ck, string $frag): ?array
    {
        $data = $this->load();
        $entry = $data["$module:$ck:$frag"] ?? null;
        return is_array($entry) ? $entry : null;
    }

    public function put(string $module, int $ck, string $frag, string $contentKeyHex, string $window, int $now): void
    {
        $data = $this->load();
        $data["$module:$ck:$frag"] = [
            'content_key_hex' => $contentKeyHex,
            'window' => $window,
            'fetched_at' => $now,
        ];
        $this->save($data);
    }
}

File: client-lib-python/vaultcode/cache.py
"""Cache locale delle content key — CIFRATA e AUTENTICATA (AES-256-GCM). §5.

File su disco = nonce(12) ‖ ciphertext‖tag. Contenuto: mappa "module:ck:frag" →
{content_key_hex, window, fetched_at}. La freschezza (TTL/grace) la decide il
Runtime, così è possibile il fallback "stale entro il grace" se il server è giù.
Formato identico alla ``Cache.php`` (la chiave di cache è la stessa derivazione).
"""

from __future__ import annotations

import hashlib
import json
import os

from . import crypto

_AAD = b"vaultcode/cache/v1"


class Cache:
    def __init__(self, path: str, key: bytes):
        self.path = path
        self.key = key

    @staticmethod
    def derive_key(install_secret: bytes) -> bytes:
        """Deriva la chiave di cache dal segreto per-installazione (deterministica,
        parità con Cache::deriveKey)."""
        return hashlib.sha256(install_secret + b"vaultcode/cache-key/v1").digest()[:32]

    def _load(self) -> dict:
        if not self.path or not os.path.isfile(self.path):
            return {}
        with open(self.path, "rb") as fh:
            blob = fh.read()
        if len(blob) < 13:
            return {}
        nonce, ct = blob[:12], blob[12:]
        try:
            data = json.loads(crypto.aes_gcm_decrypt(self.key, ct, nonce, _AAD).decode("utf-8"))
        except Exception:  # cache corrotta/manomessa → ignorata (verrà riscritta)
            return {}
        return data if isinstance(data, dict) else {}

    def _save(self, data: dict) -> None:
        if not self.path:
            return
        nonce = crypto.new_nonce()
        ct = crypto.aes_gcm_encrypt(self.key, json.dumps(data).encode("utf-8"), nonce, _AAD)
        d = os.path.dirname(self.path)
        if d and not os.path.isdir(d):
            os.makedirs(d, mode=0o700, exist_ok=True)
        with open(self.path, "wb") as fh:
            fh.write(nonce + ct)
        try:
            os.chmod(self.path, 0o600)
        except OSError:
            pass

    def get(self, module: str, ck: int, frag: str) -> dict | None:
        entry = self._load().get(f"{module}:{ck}:{frag}")
        return entry if isinstance(entry, dict) else None

    def put(self, module: str, ck: int, frag: str, content_key_hex: str, window: str, now: int) -> None:
        data = self._load()
        data[f"{module}:{ck}:{frag}"] = {
            "content_key_hex": content_key_hex, "window": window, "fetched_at": now,
        }
        self._save(data)
